bfasst/transform/spydrnet_injector.py

```diff
@@ -0,0 +1,199 @@
+            self.inject_wire_swap()
+        else:
+            return Status(TransformStatus.ERROR, "Invalid error type")
+        return Status(TransformStatus.SUCCESS)
+
+    def inject_bit_flip(self):
+        """Injects a bit flip error into the netlist"""
+        num_luts = self.pick_luts_from_netlist()
+        self.flatten_luts()
+        lut_number = randrange(num_luts)
+        lut_size = self.get_lut_size(lut_number)
+        bit_number = randrange(lut_size)
+        self.flip_bit(lut_number, bit_number)
+        self.compose_corrupt_netlist()
+
+    def pick_luts_from_netlist(self):
+        """Calculates the number of LUTs in the netlist"""
+        num_luts = 0
+        for library in self.clean_netlist.libraries:
+            for definition in library.definitions:
+                if "LUT" in definition.name.upper():
+                    num_luts += len(definition.references)
+                    self.hierarchical_luts.append(definition.references)
+        return num_luts
+
+    def flatten_luts(self):
+        """Flattens the LUTs into a single list"""
+        self.flattened_luts = [lut for sublist in self.hierarchical_luts for lut in sublist]
+
+    def get_lut_size(self, lut_number):
+        """Gets the size of the LUT init string for the given LUT"""
+        lut = self.flattened_luts[lut_number]
+        init_string = lut.data["VERILOG.Parameters"]["INIT"].upper()
+        return int(init_string.split("'")[0])
+
+    def flip_bit(self, lut_number, bit_number):
+        """Flips the bit at the given index"""
+        lut = self.flattened_luts[lut_number]
+        lut_properties = lut.data["VERILOG.Parameters"]
+
+        config_string_prefixed = lut_properties["INIT"].upper()
+        config_string_digits = config_string_prefixed.split("H")[1]
+        lut_size = self.get_lut_size(lut_number)
+        config_as_binary = bin(int(config_string_digits, 16))[2:].zfill(lut_size)
+
+        if config_as_binary[bit_number] == "1":
+            config_as_binary = (
+                config_as_binary[:bit_number] + "0" + config_as_binary[bit_number + 1 :]
+            )
+        else:
+            config_as_binary = (
+                config_as_binary[:bit_number] + "1" + config_as_binary[bit_number + 1 :]
+            )
+
+        config_as_hex = hex(int(config_as_binary, 2))
+        lut_properties["INIT"] = (
+            config_string_prefixed.split("H")[0] + "h" + str(config_as_hex).upper()[2:]
+        )
+
+    def compose_corrupt_netlist(self):
+        """Writes the netlist to the corrupted netlist path in the design"""
+        sdn.compose(self.clean_netlist, self.design.flow_paths["corrupted_netlist_path"])
+
+    def inject_wire_swap(self):
+        """Injects a wire swap error into the netlist"""
+        three_instances = self.get_random_instances(3)
+        first_instance_pins = self.get_outer_pin_inputs(three_instances[0])
+
+        selected_input = first_instance_pins[randrange(len(first_instance_pins))]
```

Comment:
Can you add more comments to these lines?  I'm not quite sure I understand what is going on here.

Reply:
Done